api/api/log_api.go

```diff
@@ -42,38 +48,33 @@ func (l *LogController) ReadLog(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	res, err := l.LogService.ReadLog(&query)
-	if err != nil {
-		log.Errorf("Error while retrieving log %v", err)
-		InternalServerError(fmt.Sprintf("Error while retrieving log for container %s: %s", query.Name, err)).WriteTo(w)
-		return
-	}
+	logLineCh := make(chan string, 1000)
+	stopCh := make(chan struct{})
 
-	// send status code and content-type
+	// set necessary headers
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
 	w.Header().Set("Content-Type", "plain/text; charset=UTF-8")
-	w.WriteHeader(http.StatusOK)
 
-	// stream the response body
-	defer res.Close()
-	buff := bufio.NewReader(res)
-	for {
-		line, readErr := buff.ReadString('\n')
-		_, writeErr := w.Write([]byte(line))
-		if writeErr != nil {
-			// connection from caller is closed
-			return
-		}
+	go func() {
```

Comment:
what happens if user open the log UI and after several seconds open another page? Should we need to listen to context change, and close all the necessary channel? So there won't be unfinished goroutine

Reply:
Currently, if the caller stop streaming the logs, we will close the stopCh and return as handled in this code: https://github.com/gojek/merlin/pull/164/files/55f9de8f7cd497d88114a3c4d5241178e4168917#diff-4d8a066a2fe8e1b33215ba5fa1142a25d9ea12586226a456f06d81b0b722dcebR64-R68

I don't think listening to context would help, but let me try it too.

Reply:
You are correct. Listening to context solves the performance issue! See this commit: https://github.com/gojek/merlin/pull/164/commits/5d4991641bd0b264aa9d3dfe733cb1b019cb51f8